node/forest_libp2p/src/blocksync/provider.rs

```diff
@@ -0,0 +1,389 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use chain::Error as ChainError;
+use forest_blocks::{Tipset, TipsetKeys};
+use forest_cid::Cid;
+use forest_encoding::tuple::*;
+use ipld_blockstore::BlockStore;
+use log::debug;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::sync::Arc;
+
+use super::{CompactedMessages, TipsetBundle};
+
+/// Blocksync request options
+pub const BLOCKS: u64 = 1;
+pub const MESSAGES: u64 = 2;
+pub const BLOCKS_MESSAGES: u64 = 3;
+
+/// The payload that gets sent to another node to request for blocks and messages. It get DagCBOR serialized before sending over the wire.
+#[derive(Clone, Debug, PartialEq, Serialize_tuple, Deserialize_tuple)]
+pub struct BlockSyncRequest {
+    /// The tipset to start sync from
+    pub start: Vec<Cid>,
+    /// The amount of epochs to sync by
+    pub request_len: u64,
+    /// 1 = Block only, 2 = Messages only, 3 = Blocks and Messages
+    pub options: u64,
+}
+
+impl BlockSyncRequest {
+    pub fn include_blocks(&self) -> bool {
+        return self.options == BLOCKS || self.options == BLOCKS_MESSAGES;
+    }
+
+    pub fn include_messages(&self) -> bool {
+        return self.options == MESSAGES || self.options == BLOCKS_MESSAGES;
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum BlockSyncResponseStatus {
+    // All is well.
+    Success = 1,
+    // We could not fetch all blocks requested (but at least we returned
+    // the `Head` requested). Not considered an error.
+    PartialResponse = 101,
+    // Request.Start not found.
+    BlockNotFound = 201,
+    // Requester is making too many requests.
+    GoAway = 202,
+    // Internal error occured.
+    InternalError = 203,
+    // Request was bad
+    BadRequest = 204,
+}
```

Comment:
yeah, my bad, tested the serialization and it's indeed wrong. 
So I was wondering if in rust there is a way to represent the integers in enum, and be able to serialize them properly. 

I can think of two ways:

```
struct Status(u64);

impl Struct {
  pub fn set_success() {
   // check me on this
    self.0 = 500;
  }

  pub fn get_status() {
    self.0
  }
  
}
```

or use some crate like https://docs.rs/num_enum/0.5.1/num_enum/

Anyways, switched from `enum` to `const`s

Reply:
Yeah so there are a few ways to match using [serde repr](https://github.com/dtolnay/serde-repr) which isn't as ideal in this case because it'll only deserialize if values match, or you can just manually implement deserialize through an integer and just have an enum for Other codes for when it doesn't match.

Here is the change I am suggesting as a commit, for consideration: https://github.com/ChainSafe/forest/commit/61e857033dbef88b600a76bfd5de64034bd34e8d. Using an integer value with constants is completely fine as well.

Reply:
nice! Replaced with your suggestion.